diff --git a/device/generic/goldfish/hals/hwc3/Device.cpp b/device/generic/goldfish/hals/hwc3/Device.cpp
--- a/device/generic/goldfish/hals/hwc3/Device.cpp
+++ b/device/generic/goldfish/hals/hwc3/Device.cpp
@@ -93,3 +93,18 @@
     std::unique_lock<std::mutex> lock(mMutex);
 
+    if (mComposer == nullptr &&
+        ::android::base::GetProperty("ro.vendor.hwcomposer.mode", "") == "guest") {
+        DEBUG_LOG("%s: using Kiki GuestFrameComposer", __FUNCTION__);
+        mComposer = std::make_unique<GuestFrameComposer>();
+        if (!mComposer) {
+            ALOGE("%s failed to allocate GuestFrameComposer", __FUNCTION__);
+            return HWC3::Error::NoResources;
+        }
+        HWC3::Error error = mComposer->init();
+        if (error != HWC3::Error::None) {
+            ALOGE("%s failed to init GuestFrameComposer", __FUNCTION__);
+            return error;
+        }
+    }
+
     if (mComposer == nullptr) {
